[CARBONDATA-2016] Exception displays while executing compaction with alter query

anubhav100 opened the pull request:
**Root Cause**
    When we apply the alter table command to add column with default value it is storing it as long object,it is wrongly written in restructure util we should get the value as the same type
     as that of datatype of columnschema in restructure util earlier in master branch in restuctureutil class if  our data type is long or short or int we are always returning back a long object which is wong due to same reason compaction was failing
    if it was applied after alter table add columns command with default value  because in sortdatarows there was mismatch between data type and its corresponding value

**Testing:**
1.mvn clean install is passing
2.added new test case for same

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2703/

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1470/

anubhav100 commented:
retest this please

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1473/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2706/

ravipesala commented:
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2838/

anubhav100 commented:
retest this please

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2709/

anubhav100 commented:
retest this please

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1477/

anubhav100 commented:
retest this please

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1478/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2711/

anubhav100 commented:
retest this please

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2712/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1479/

jackylk commented on processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/SortDataRows.java:
```diff
@@ -270,7 +270,7 @@ private void writeDataToFile(Object[][] recordHolderList, int entryCountLocal, F
               stream.writeShort((Short) value);
             } else if (dataType == DataTypes.INT) {
-              stream.writeInt((Integer) value);
+              stream.writeInt(Integer.parseInt(value.toString()));
```
Can you post the error message and describe what is the root cause of this bug?

anubhav100 replied:
https://issues.apache.org/jira/browse/CARBONDATA-2016

anubhav100 replied:
here is exception trace 18/01/10 16:17:12 ERROR CompactionResultSortProcessor: [Executor task launch worker-36][partitionID:customer1;queryID:15798380253871] Compaction failed: java.lang.Long cannot be cast to java.lang.Integer
java.lang.ClassCastException: java.lang.Long cannot be cast to java.lang.Integer
at org.apache.carbondata.processing.sort.sortdata.SortDataRows.writeDataToFile(SortDataRows.java:273)
at org.apache.carbondata.processing.sort.sortdata.SortDataRows.startSorting(SortDataRows.java:214)
at org.apache.carbondata.processing.merger.CompactionResultSortProcessor.processResult(CompactionResultSortProcessor.java:226)
at org.apache.carbondata.processing.merger.CompactionResultSortProcessor.execute(CompactionResultSortProcessor.java:159)
at org.apache.carbondata.spark.rdd.CarbonMergerRDD$$anon$1.<init>(CarbonMergerRDD.scala:234)
at org.apache.carbondata.spark.rdd.CarbonMergerRDD.internalCompute(CarbonMergerRDD.scala:81)

jackylk replied:
But what is the root cause of this bug? I think your change (convert to string and then int) is like hiding this bug.

anubhav100 replied:
@jackylyk ok i get your point you mean we should handle it at the base root level ok I am looking into it

anubhav100 commented:
@jackylk  

**here is the root reason**

**when we apply the alter table command with default value i;e
       "alter table customer1 add columns (intfield int) TBLPROPERTIES ('DEFAULT.VALUE.intfield'='10')")

that value is always return as the long object in restructure util thats why compaction failed because the column value for newly added column is long but the column type is int**


it is wrongly written in restructure util  this code is wrong

  public static Object getMeasureDefaultValue(ColumnSchema columnSchema, byte[] defaultValue) {
    Object measureDefaultValue = null;
    if (!isDefaultValueNull(defaultValue)) {
      String value = null;
      DataType dataType = columnSchema.getDataType();
      if (dataType == DataTypes.SHORT || dataType == DataTypes.INT || dataType == DataTypes.LONG) {
        value = new String(defaultValue, Charset.forName(CarbonCommonConstants.DEFAULT_CHARSET));
        measureDefaultValue = Long.parseLong(value);
      } else if (DataTypes.isDecimal(dataType)) {
        BigDecimal decimal = DataTypeUtil.byteToBigDecimal(defaultValue);
        if (columnSchema.getScale() > decimal.scale()) {
          decimal = decimal.setScale(columnSchema.getScale(), RoundingMode.HALF_UP);
        }
        measureDefaultValue = decimal;
      } else {
        value = new String(defaultValue, Charset.forName(CarbonCommonConstants.DEFAULT_CHARSET));
        Double parsedValue = Double.valueOf(value);
        if (!Double.isInfinite(parsedValue) && !Double.isNaN(parsedValue)) {
          measureDefaultValue = parsedValue;
        }
      }
    }
    return measureDefaultValue;
  }

this code is always returning a long value objetc that is wrong

i corrected it as 

  */
  public static Object getMeasureDefaultValue(ColumnSchema columnSchema, byte[] defaultValue) {
    Object measureDefaultValue = null;
    if (!isDefaultValueNull(defaultValue)) {
      String value = null;
      DataType dataType = columnSchema.getDataType();
      if (dataType == DataTypes.SHORT) {
        value = new String(defaultValue, Charset.forName(CarbonCommonConstants.DEFAULT_CHARSET));
        measureDefaultValue = new Short(value);
      }
     else if (dataType == DataTypes.LONG) {
        value = new String(defaultValue, Charset.forName(CarbonCommonConstants.DEFAULT_CHARSET));
        measureDefaultValue = Long.parseLong(value);
      }
     else if (dataType == DataTypes.INT) {
        value = new String(defaultValue, Charset.forName(CarbonCommonConstants.DEFAULT_CHARSET));
        measureDefaultValue = Integer.parseInt(value);
      }
      else if (DataTypes.isDecimal(dataType)) {
        BigDecimal decimal = DataTypeUtil.byteToBigDecimal(defaultValue);
        if (columnSchema.getScale() > decimal.scale()) {
          decimal = decimal.setScale(columnSchema.getScale(), RoundingMode.HALF_UP);
        }
        measureDefaultValue = decimal;
      } else {
        value = new String(defaultValue, Charset.forName(CarbonCommonConstants.DEFAULT_CHARSET));
        Double parsedValue = Double.valueOf(value);
        if (!Double.isInfinite(parsedValue) && !Double.isNaN(parsedValue)) {
          measureDefaultValue = parsedValue;
        }
      }
    }
    return measureDefaultValue;
  }

please review i corrected it

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1662/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2896/

anubhav100 commented:
retest this please

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1672/

anubhav100 commented:
retest this please

anubhav100 commented:
retest this please

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2904/

anubhav100 commented:
retest this please

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1676/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2909/

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1681/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2914/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1688/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2921/

ravipesala commented:
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2963/

anubhav100 commented:
retest sdv please

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2970/

anubhav100 commented:
@jackylk please review

jackylk commented:
LGTM

jackylk commented:
please squash and push again

anubhav100 commented:
@jackylk  i accidently deleted the commits of this branch when squashing the commit i raised the new pull request with same changes please review #1839

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3008/
